File: 23-MergekSortedLists/23-MergekSortedLists.py
# Last updated: 4/23/2026, 3:42:26 PM
1# Definition for singly-linked list.
2# class ListNode:
3#     def __init__(self, val=0, next=None):
4#         self.val = val
5#         self.next = next
6class Solution:
7    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
8        pq = []
9        count = 0
10        for i in lists:
11            if i:
12                heapq.heappush(pq, (i.val, count, i))
13                count += 1
14        if not pq: return None
15        head = heapq.heappop(pq)[2]
16        if head.next: heapq.heappush(pq, (head.next.val, count, head.next))
17        count += 1
18        tail = head
19        while pq:
20            tail.next = heapq.heappop(pq)[2]
21            tail = tail.next
22            if tail.next: heapq.heappush(pq, (tail.next.val, count, tail.next))
23            count += 1
24        tail.next = None
25        return head

File: 23-MergekSortedLists/23-MergekSortedLists.cpp
// Last updated: 4/23/2026, 3:23:45 PM
1/**
2 * Definition for singly-linked list.
3 * struct ListNode {
4 *     int val;
5 *     ListNode *next;
6 *     ListNode() : val(0), next(nullptr) {}
7 *     ListNode(int x) : val(x), next(nullptr) {}
8 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
9 * };
10 */
11class Solution {
12public:
13    ListNode* mergeKLists(vector<ListNode*>& lists) {
14        auto comp = [](const ListNode* a, const ListNode* b) { return a->val > b->val; };
15        priority_queue<ListNode*, vector<ListNode*>, decltype(comp)> pq(comp);
16        for (auto& i : lists) if (i) pq.push(i);
17
18        if (pq.empty()) return nullptr;
19        
20        ListNode* head = pq.top();
21        pq.pop();
22        if (head->next) pq.push(head->next);
23        ListNode* tail = head;
24        while (!pq.empty()) {
25            tail->next = pq.top();
26            pq.pop();
27            tail = tail->next;
28            if (tail->next) pq.push(tail->next);
29        }
30        tail->next = nullptr;
31        return head;
32    }
33};